en/guide/message_signing.md

```diff
@@ -1,121 +1,150 @@
+## Accepting Unsigned Packets
 
-The intention with the link IDs is that each channel of communication between an autopilot and a GCS uses a different link ID. There is no requirement that the same link ID be used in both directions however.
+The rules for accepting *unsigned* packets will be implementation specific, but could be based on a combination of a parameter setting, transport type, message type, (in)compatibility flags etc.
 
-For C implementations the obvious mechanism is to use the MAVLink channel number as the link ID. That works well for an autopilot, but runs into an issue for a GCS implementation. The issue is that a user may launch multiple GCS instances talking to the same autopilot via different communication links (such as two radios, or USB and a radio). These multiple GCS instances will not be aware of each other, and so may choose the same link ID. If that happens then a large number of correctly signed packets will be rejected by the autopilot as they will have timestamps that are older than the timestamp received for the same stream tuple on the other communication link.
+> **Note** MAVLink libraries should provide a mechanism that allows an implementation to make this kind of decision.
 
-The solution adopted for MAVProxy is shown below:
+A few rules that may be useful for some systems are:
+* All unsigned packets accepted based on a system-specific parameter.
+* All unsigned packets accepted if the connection is over a "secure channel" (e.g. local USB cable or local wired Ethernet cable).
+* `RADIO_STATUS` packets are always accepted without signing (to make life easier for telemetry radios).
+* All unsigned packets accepted when in an "unsigned mode" (perhaps triggered by a hardware button pressed on boot).
+* MAVLink 1 messages may always be accepted or always rejected, depending on the system requirements.
+* All other packets will be rejected if not signed!
 
-```
-if (msg.get_signed() and
-	self.mav.signing.link_id == 0 and
-	msg.get_link_id() != 0 and
-	self.target_system == msg.get_srcSystem() and
-	self.target_component == msg.get_srcComponent()):
-	# change to link_id from incoming packet
-	self.mav.signing.link_id = msg.get_link_id()
-```
+<!-- 
+How are we handling case of signed packets that meet the criteria for accepting unsigned packets but signature is incorrect or the timestamp rules are not met. E.g. for lost copter finding or re-keying? AFAIK no support in the C library for handling this?
+-->
+
+
+## Secret Key Management {#secret_key}
+
+A secret key is 32 bytes of binary data that are used to create message signatures that can be verified by other holders of the key.
+The key should be created on one system in the network (often a GCS) and shared to other trusted devices via secure channels.
+All systems on a link will use the *same* key. Different links may use different keys.
```

Comment:
Is this correct. I'm assuming so, but it may be that I don't understand "link, channel, stream" properly.

Reply:
@WickedShell I wrote this because the C implementation stores a single key for each link.
1. I can see that there is nothing to stop having unique keys for each system, but 
  - what is the benefit?
1. How you would you then support multicast - since some of the systems on the network might not understand the key. 

Am I missing your point?

As an aside, How do you define "link"?
I tend to see it as a connection to a distinct Mavlink network accessible to a MAVLink system via a single port/channel (such that all communicating systems can reasonably share a common understanding of the timestamp).




Did any of that make sense or am I completely missing the point?

Reply:
> because the C implementation stores a single key for each link

Which is fair, but it's not a protocol limitation, it's simply a limitation of a parser. (The Clojure one can juggle multiple keys for example).

> 1. I can see that there is nothing to stop having unique keys for each system, but
>     - what is the benefit?

Either common link forwarding, or partitioned access comes to mind. You might have a vehicle which is being controlled by one source, and using the same transport link to the vehicle and packet forwarding a payload that you want to only allow command and control for each of these to come from different applications. The other one this gives you is a very dirty workaround to the fact that not all MAVLink messages can be targeted. None of these are deployed use cases as far as I know, but they are viable at the moment.

> How you would you then support multicast - since some of the systems on the network might not understand the key.

It relates to all of the same points as in the above use cases. I don't think there are a ton of reasons to do multicast for this, but I've not had a lot to do with this kind of network communications so I could be missing a reason to/not to do this.

Reply:
Thanks. I've added a clarification that the C library only allows one, but it is not a limitation of the protocol